src/rust/engine/nailgun/src/server.rs

```diff
@@ -0,0 +1,355 @@
+    // acquires `write`.
+    let ongoing_connections = Arc::new(RwLock::new(()));
+
+    let result = loop {
+      let tcp_stream = match future::select(listener.accept().boxed(), should_exit).await {
+        future::Either::Left((Ok((tcp_stream, _addr)), s_e)) => {
+          // Got a connection.
+          should_exit = s_e;
+          tcp_stream
+        }
+        future::Either::Left((Err(e), _)) => {
+          break Err(format!("Server failed to accept connections: {}", e));
+        }
+        future::Either::Right((_, _)) => {
+          break Ok(());
+        }
+      };
+
+      debug!("Accepted connection: {:?}", tcp_stream);
+
+      // There is a slightly delicate dance here: we wait for a connection to have acquired the
+      // ongoing connections lock before proceeding to the next iteration of the loop. This
+      // prevents us from observing an empty lock and exiting before the connection has actually
+      // acquired it. Unfortunately we cannot acquire the lock in this thread and then send the
+      // guard to the other thread due to its lifetime bounds.
+      let connection_started = Arc::new(Notify::new());
+      let _join = executor.spawn({
+        let config = config.clone();
+        let nail = nail.clone();
+        let connection_started = connection_started.clone();
+        let ongoing_connections = ongoing_connections.clone();
+        async move {
+          let ongoing_connection_guard = ongoing_connections.read().await;
+          connection_started.notify();
+          let result = nails::server_handle_connection(config, nail, tcp_stream).await;
+          std::mem::drop(ongoing_connection_guard);
+          result
+        }
+      });
+      connection_started.notified().await;
+    };
+
+    // Before exiting, acquire write access on the ongoing_connections lock to prove that all
+    // connections have completed.
+    debug!("Server waiting for connections to complete...");
+    let _ = ongoing_connections.write().await;
+    debug!("All connections completed.");
+    result
+  }
+
+  ///
+  /// The port that the server is listening on.
+  ///
+  pub fn port(&self) -> u16 {
+    self.port
+  }
+
+  ///
+  /// Returns a Future that will shut down the server by:
+  /// 1. stopping accepting new connections
+  /// 2. waiting for all ongoing connections to have completed
+  ///
+  pub async fn shutdown(self) -> Result<(), String> {
+    // If we fail to send the exit signal, it's because the task is already shut down.
+    let _ = self.exit_sender.send(());
+    self
+      .exited_receiver
+      .await
+      .map_err(|_| "Server exited uncleanly.".to_owned())?
+  }
+}
+
+pub struct RawFdExecution {
+  pub cmd: execution::Command,
+  pub stdin_fd: RawFd,
+  pub stdout_fd: RawFd,
+  pub stderr_fd: RawFd,
+}
+
+///
+/// Implementations of this trait should _not_ close the input file handles, and should let the
```

Comment:
I might add a TODO to figure out how to ensure this condition with lifetimes or types somehow.